UniStuttgart-VISUS/megamol
# Allow to Register Lambdas as Lua Callbacks

* Introduce LuaCallbacksCollection resource - wrapper for lambda functions as Lua callbacks
* Add corresponding Add/Remove/Clear methods to LuaAPI - registers and removes Lua lambda callbacks on the fly
* Implement Lua Config parsing callbacks as lambdas in `CLIConfigParsing.cpp`
* Implement Lua Callbacks that wrap frontend resources (screenshot, fbo resize, fullscreen, vsync, gui state/show/scale) as lambdas in `Lua_Service_Wrapper.cpp`
* Implement Lua Callbacks that manipulate the MegaMolGraph (create module, get param value, ...) as lambdas in `Lua_Service_Wrapper.cpp`
* This way Lua callbacks are only available in Lua when they are safe to use: when the resources used by the lambdas are initialized and ready to use
* Add `--guishow` and `--guiscale` CLI options

## Summary of Changes
Provide abstraction to register C++ lambdas from frontend or graph modules as Lua callbacks. Implement most existing Lua callbacks with this structure.

## References and Context
Leads to easier Lua callback writing and prototyping. Frontend services can provide Lua callbacks on their own, without the need to touch the central LuaAPI.cpp file. Even Graph modules can access the `RegisterLuaCallbacks(megamol::frontend_resources::LuaCallbacksCollection const& callbacks)` frontend resource function and register Lua callbacks of their own.

The type of the lambda to register as Lua callback needs to be known at compile time. However, actual registration of the callbacks with Lua happens at runtime. The template machinery of the `LuaCallbacksCollection` class provides the lambda input arguments with the expected inputs from the Lua stack, forwards errors returned via `LuaError{"description"}` to Lua and pushes values returned by the lambda to the Lua stack. All in all, the registered lambdas behave as normal Lua callbacks, but with an easier interface and better usability.

## Test Instructions
Load complicated Config and Project files and see whether something fails that shoud not.
